thrust/system/cuda/detail/util.h

```diff
@@ -95,23 +96,33 @@ cudaError_t
 synchronize_stream(execution_policy<Derived> &policy)
 {
   cudaError_t result;
-  if (THRUST_IS_HOST_CODE) {
-    #if THRUST_INCLUDE_HOST_CODE
-      cudaStreamSynchronize(stream(policy));
-      result = cudaGetLastError();
-    #endif
-  } else {
-    #if THRUST_INCLUDE_DEVICE_CODE
-      #if __THRUST_HAS_CUDART__
-        THRUST_UNUSED_VAR(policy);
-        cub::detail::device_synchronize();
-        result = cudaGetLastError();
-      #else
-        THRUST_UNUSED_VAR(policy);
-        result = cudaSuccess;
-      #endif
-    #endif
-  }
+
+  // Can't use #if inside NV_IF_TARGET, use a temp macro to hoist the device
+  // instructions out of the target logic.
+#if __THRUST_HAS_CUDART__
+
+#define THRUST_TEMP_DEVICE_CODE \
+  THRUST_UNUSED_VAR(policy); \
+  cub::detail::device_synchronize(); \
+  result = cudaGetLastError()
+
+#else
+
+#define THRUST_TEMP_DEVICE_CODE \
+  THRUST_UNUSED_VAR(policy); \
```

Comment:
I'm a bit concerned about returning `cudaSuccess` when the code is unable to perform synchronization. Should we return an error in this case instead? Also, is there any chance that we can use `cub::SyncStream` here?

Reply:
Agreed -- updated.